### PR TITLE
Make default InferType accept other dtypes than Float32 fixes #2260

### LINKED ISSUE #2260: InferType for simple op (for types other than Float32)
InferType fails for NDarray with element type other than `Float32`. In https://github.com/dmlc/mxnet/blob/f5f18b2297c581bb53bdaf1e407151f91336675d/include/mxnet/operator.h#L265, while https://github.com/dmlc/mxnet/blob/e63f7a6d0065f6b2d5fae71fd5d5ff70786b0d20/src/operator/elementwise_binary_op-inl.h#L22 seems to indicate that it supports all DTypes that mshadow supports.

### PR DESCRIPTION
**Pull request by vchuravy**
InferType now requires for dtypes to be homogenous instead of being Float32.

Ref #2260

### CODE REVIEW

**Comment by piiswrong** [comment]
The default infertype in base operator only support float32 because most ops only support float32. Override simple op's infertype instead

**Comment by vchuravy** [comment]
I think we should push forward to make most Operator support other types. Right now `Float16` support is crippled by the fact that some common operators are not supported. 
I am working right now on making pooling support more types and the next on my list is `LeakyReLU`.

**Comment by vchuravy** [comment]
So from what I can tell the following operators currently don't support anything than `real_t` e.g. `Float32`.

- [ ] `regression_output`
- [ ] `reshape`
- [ ] `swapaxis`
- [ ] `softmax_activation`
- [ ] `dropout`
- [ ] `pooling` #2280 
- [ ] `lrn`
- [ ] `roi_pooling`
- [ ] `elementwise_sum`
- [ ] `crop`
- [ ] `slice_channel`
- [ ] `leaky_relu`
- [ ] `batch_norm`
- [ ] `upsampling`
- [ ] `deconvolution`
- [ ] `concat`
- [ ] `block_grad`

For my own work I will have to add support for `batch_norm`, `pooling`, `concat`, `upsampling`, `deconvolution`.

**Comment by piiswrong** [comment]
The problem is this change can break backward compatibility. Users may have their own ops that only support float32

**Comment by vchuravy** [comment]
I would argue thought that this is less of a problem, since they wouldn't be calling their ops with input types that are different from `Float32`.

The potential issues I see are:
- Custom ops that rely on inhomogeneous input types, but those would have needed a custom `InferType` already.
- All input types are unspecified `-1`, but that should have been an error from the beginning and also should not happen.
- If an op is called that doesn't support anything, but `Float32`, we might get an non-descriptive error. To mitigate this I could add InferTypes methods to all the ops listed above that specify restrict the input type to Float32.

**Comment by vchuravy** [comment]
Let's go for now with specific InferTypes in operators that support other DTypes.
